## Segmenting and Clustering Neighborhoods in Toronto (Part 2)

This notebook consists of answer for Part 1 and 2 of the Assignment.

## Part 1

In [1]:
!pip install bs4
print('\nBeautifulSoup package installation success.')


BeautifulSoup package installation success.


Import libraries

In [2]:
from bs4 import BeautifulSoup 
import requests
import pandas as pd

The following url contains html tables with list of postal codes in Canada where the first letter is M. Postal codes beginning with M are located within the city of Toronto.

In [3]:
url = "https://en.wikipedia.org/wiki/List_of_postal_codes_of_Canada:_M"

Read the HTML tables inside the Wikipedia page.

In [4]:
table_data = pd.read_html(url, flavor='bs4')
print(f'Total tables: {len(table_data)}')

Total tables: 3


In [5]:
table_data[0].head()

0                1  \
0                                    M1ANot assigned  M2ANot assigned   
1                    M1BScarborough(Malvern / Rouge)  M2BNot assigned   
2  M1CScarborough(Rouge Hill / Port Union / Highl...  M2CNot assigned   
3  M1EScarborough(Guildwood / Morningside / West ...  M2ENot assigned   
4                             M1GScarborough(Woburn)  M2GNot assigned   

                                                2  \
0                        M3ANorth York(Parkwoods)   
1                   M3BNorth York(Don Mills)North   
2  M3CNorth York(Don Mills)South(Flemingdon Park)   
3                                 M3ENot assigned   
4                                 M3GNot assigned   

                                                3  \
0                 M4ANorth York(Victoria Village)   
1  M4BEast York(Parkview Hill / Woodbine Gardens)   
2                  M4CEast York(Woodbine Heights)   
3                    M4EEast Toronto(The Beaches)   
4                           M4GEast York(Leaside)   

                                                 4  \
0  M5ADowntown Toronto(Regent Park / Harbourfront)   
1    M5BDowntown Toronto(Garden District, Ryerson)   
2              M5CDowntown Toronto(St. James Town)   
3                 M5EDowntown Toronto(Berczy Park)   
4          M5GDowntown Toronto(Central Bay Street)   

                                                  5  \
0  M6ANorth York(Lawrence Manor / Lawrence Heights)   
1                          M6BNorth York(Glencairn)   
2                       M6CYork(Humewood-Cedarvale)   
3                      M6EYork(Caledonia-Fairbanks)   
4                     M6GDowntown Toronto(Christie)   

                                                6                7  \
0  M7AQueen's Park(Ontario Provincial Government)  M8ANot assigned   
1                                 M7BNot assigned  M8BNot assigned   
2                                 M7CNot assigned  M8CNot assigned   
3                                 M7ENot assigned  M8ENot assigned   
4                                 M7GNot assigned  M8GNot assigned   

                                                   8  
0                     M9AEtobicoke(Islington Avenue)  
1  M9BEtobicoke(West Deane Park / Princess Garden...  
2  M9CEtobicoke(Eringate / Bloordale Gardens / Ol...  
3                                    M9ENot assigned  
4                                    M9GNot assigned

We want to capture Postal codes, Borough and Neighborhood data from the first table.

Use BeautifulSoup package to transform the Wikipedia data into dataframe.

In [6]:
url_data  = requests.get(url).text
soup = BeautifulSoup(url_data,"html5lib")
tables = soup.find_all('table')
table_data = pd.DataFrame(columns=['PostalCode', 'Borough', 'Neighborhood'])

for row in tables[0].tbody.find_all('td'):
    if row.span.text == 'Not assigned':
        pass
    else:
        postalcode = row.p.text[:3] 
        borough = (row.span.text).split('(')[0]
        neighborhood = (((((row.span.text).split('(')[1]).strip(')')).replace(' /',',')).replace(')',' ')).strip(' ')
        table_data = table_data.append({"PostalCode" : postalcode, "Borough" : borough, "Neighborhood" : neighborhood}, ignore_index=True)

table_data

PostalCode                                            Borough  \
0          M3A                                         North York   
1          M4A                                         North York   
2          M5A                                   Downtown Toronto   
3          M6A                                         North York   
4          M7A                                       Queen's Park   
..         ...                                                ...   
98         M8X                                          Etobicoke   
99         M4Y                                   Downtown Toronto   
100        M7Y  East TorontoBusiness reply mail Processing Cen...   
101        M8Y                                          Etobicoke   
102        M8Z                                          Etobicoke   

                                          Neighborhood  
0                                            Parkwoods  
1                                     Victoria Village  
2                            Regent Park, Harbourfront  
3                     Lawrence Manor, Lawrence Heights  
4                        Ontario Provincial Government  
..                                                 ...  
98       The Kingsway, Montgomery Road, Old Mill North  
99                                Church and Wellesley  
100                                     Enclave of M4L  
101  Old Mill South, King's Mill Park, Sunnylea, Hu...  
102  Mimico NW, The Queensway West, South of Bloor,...  

[103 rows x 3 columns]

In [7]:
table_df = pd.DataFrame(table_data)
table_df['Borough'] = table_df['Borough'].replace({'Downtown TorontoStn A PO Boxes25 The Esplanade':'Downtown Toronto Stn A',
                                             'East TorontoBusiness reply mail Processing Centre969 Eastern':'East Toronto Business',
                                             'EtobicokeNorthwest':'Etobicoke Northwest','East YorkEast Toronto':'East York/East Toronto',
                                             'MississaugaCanada Post Gateway Processing Centre':'Mississauga'})
table_df

PostalCode                Borough  \
0          M3A             North York   
1          M4A             North York   
2          M5A       Downtown Toronto   
3          M6A             North York   
4          M7A           Queen's Park   
..         ...                    ...   
98         M8X              Etobicoke   
99         M4Y       Downtown Toronto   
100        M7Y  East Toronto Business   
101        M8Y              Etobicoke   
102        M8Z              Etobicoke   

                                          Neighborhood  
0                                            Parkwoods  
1                                     Victoria Village  
2                            Regent Park, Harbourfront  
3                     Lawrence Manor, Lawrence Heights  
4                        Ontario Provincial Government  
..                                                 ...  
98       The Kingsway, Montgomery Road, Old Mill North  
99                                Church and Wellesley  
100                                     Enclave of M4L  
101  Old Mill South, King's Mill Park, Sunnylea, Hu...  
102  Mimico NW, The Queensway West, South of Bloor,...  

[103 rows x 3 columns]

Make sure "Not assigned" neighborhood is not inside dataframe.

In [8]:
table_df.Neighborhood.str.count('Not assigned').sum()

0

In [9]:
len(table_df['PostalCode'].unique())

103

There are 103 unique postal code inside the dataframe.

In [10]:
table_df.shape

(103, 3)

### Answer: There are 103 rows inside the dataframe.

---

## Part 2

In [11]:
!pip install geocoder
print('\nGeocoder package installation success.')


Geocoder package installation success.


Get the latitude and longitude coordinates of each neighborhood in Toronto.

In [12]:
import geocoder

geo_coords = None
postal_code = 'M5G'
count = 0

while (geo_coords is None and count < 2500):
    geo = geocoder.google('{}, Toronto, Ontario'.format(postal_code))
    count = count + 1
    geo_coords = geo.latlng
    
if (geo_coords is not None):
    latitude = geo_coords[0]
    longitude = geo_coords[1]
    print(latitude, longitude)
else:
    print('Try ' + str(count) + ' times to get the geographical coordinates through geocoder but not success.')

Try 2500 times to get the geographical coordinates through geocoder but not success.


Not able to get the geographical coordinates of the neighborhoods using the Geocoder package. Get the geographical coordinates for each postal code from csv file.

In [13]:
coordinate_data = pd.read_csv('https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/IBMDeveloperSkillsNetwork-DS0701EN-SkillsNetwork/labs_v1/Geospatial_Coordinates.csv')
coordinate_data.columns = ['PostalCode', 'Latitude', 'Longitude']
coordinate_data.sort_values(by=['PostalCode'], ascending=True, inplace=True)
coordinate_data

PostalCode   Latitude  Longitude
0          M1B  43.806686 -79.194353
1          M1C  43.784535 -79.160497
2          M1E  43.763573 -79.188711
3          M1G  43.770992 -79.216917
4          M1H  43.773136 -79.239476
..         ...        ...        ...
98         M9N  43.706876 -79.518188
99         M9P  43.696319 -79.532242
100        M9R  43.688905 -79.554724
101        M9V  43.739416 -79.588437
102        M9W  43.706748 -79.594054

[103 rows x 3 columns]

Sorting ascending on table_contents data.

In [14]:
table_df.sort_values(by=['PostalCode'], ascending=True, inplace=True, ignore_index=True)
table_df

PostalCode              Borough  \
0          M1B          Scarborough   
1          M1C          Scarborough   
2          M1E          Scarborough   
3          M1G          Scarborough   
4          M1H          Scarborough   
..         ...                  ...   
98         M9N                 York   
99         M9P            Etobicoke   
100        M9R            Etobicoke   
101        M9V            Etobicoke   
102        M9W  Etobicoke Northwest   

                                          Neighborhood  
0                                       Malvern, Rouge  
1               Rouge Hill, Port Union, Highland Creek  
2                    Guildwood, Morningside, West Hill  
3                                               Woburn  
4                                            Cedarbrae  
..                                                 ...  
98                                              Weston  
99                                           Westmount  
100  Kingsview Village, St. Phillips, Martin Grove ...  
101  South Steeles, Silverstone, Humbergate, Jamest...  
102  Clairville, Humberwood, Woodbine Downs, West H...  

[103 rows x 3 columns]

Merge the table_df and coordinate_data dataframe together.

In [15]:
toronto_df = pd.merge(table_df, coordinate_data, how="left", on=["PostalCode"])
toronto_df

PostalCode              Borough  \
0          M1B          Scarborough   
1          M1C          Scarborough   
2          M1E          Scarborough   
3          M1G          Scarborough   
4          M1H          Scarborough   
..         ...                  ...   
98         M9N                 York   
99         M9P            Etobicoke   
100        M9R            Etobicoke   
101        M9V            Etobicoke   
102        M9W  Etobicoke Northwest   

                                          Neighborhood   Latitude  Longitude  
0                                       Malvern, Rouge  43.806686 -79.194353  
1               Rouge Hill, Port Union, Highland Creek  43.784535 -79.160497  
2                    Guildwood, Morningside, West Hill  43.763573 -79.188711  
3                                               Woburn  43.770992 -79.216917  
4                                            Cedarbrae  43.773136 -79.239476  
..                                                 ...        ...        ...  
98                                              Weston  43.706876 -79.518188  
99                                           Westmount  43.696319 -79.532242  
100  Kingsview Village, St. Phillips, Martin Grove ...  43.688905 -79.554724  
101  South Steeles, Silverstone, Humbergate, Jamest...  43.739416 -79.588437  
102  Clairville, Humberwood, Woodbine Downs, West H...  43.706748 -79.594054  

[103 rows x 5 columns]

In [16]:
toronto_df.shape

(103, 5)

We get 103 records as we execute the left merge method on 'PosalCode' key.

---